### ISSUE TITLE
Rename ``FirstChanceExceptionStatisticsTelemetryModule`` to ``FirstChanceExceptionMetricsTelemetryModule`

### ISSUE THREAD

**Issue by AlexBulankou** [Microsoft/ApplicationInsights-dotnet-server#491]
Suggestion: rename ``FirstChanceExceptionStatisticsTelemetryModule`` to ``FirstChanceExceptionMetricsTelemetryModule``. ``Statistics`` has implication of some statistical analysis that is happening as part of collection, and this is not the case. I think ``Metrics`` is more appropriate reference to what it is emitting.

**Comment by SergeyKanzhelev** [comment]
We are doing both - metrics and examples based on statistical analysis =)

I was thinking to move the logic of pre-aggregation to the processor that is making all other pre-aggregations and generalize "examples" collection logic to other types. I'll rename the issue.

**Comment by SergeyKanzhelev** [comment]
Do you think we need to do it for 2.4 stable?

**Comment by AlexBulankou** [comment]
@SergeyKanzhelev No, but I think if we release it with this name now, next opportunity to make a breaking change (since this public API) will be 3.0, not 2.5. So either 2.4 or 3.0.

**Comment by ohadschn** [comment]
More importantly, this module needs to be documented *somewhere*. There are literally less than 10 results when Binging/Googling `FirstChanceExceptionStatisticsTelemetryModule`, none of which actually explaining what it does. 

In the meantime, I'll paste what the [source](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/8d88ec1787442c6299c688f743d26380e705209f/Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs#L19) says for the benefit of web searchers getting here:

> The module subscribed to AppDomain.CurrentDomain.FirstChanceException to send exceptions statistics to ApplicationInsights.

**Comment by SergeyKanzhelev** [comment]
It is a part of this story: https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/212 Commented as we rationalizing the story now. I agree, we need to document it. I was planning to do it. Now that you mentioned, I'll try to squeeze it into the next week.

It was enabled on some of our services so it's quite stable. If you'd enable it in your environment - give feedback! Especially important feedback on the balance of "usefullness" vs. "noisiness"

**Comment by ohadschn** [comment]
I actually did enable it (without really realizing what it does, I'm rash like that), and I'm going to have to lean towards "noisiness". I could make the argument that most first chance exceptions are not interesting at all, much less when they clutter up the super-important exceptions space:

 1. In external code, you have nothing to do about them and they are basically pure noise (not to point fingers, but `PerfCounterCollector` is quite the offender ;])
 1. In one's own (properly written) code, they would mostly be valid [vexing and exogenous exceptions](https://blogs.msdn.microsoft.com/ericlippert/2008/09/10/vexing-exceptions/).
 1. Exceptions that do slip through should be caught by either custom `TrackException` code or some unhandled exception module.

Most people won't realize they need to filter by `_MS.Example`, and be swamped with irrelevant exceptions. Maybe a different `firstChanceException` schema object would be better. Also, these exceptions currently appear just like any "real" exception in the "Failures" blade, skewing the statistics along the way (e.g. "Total Server Exceptions by Problem ID").

BTW having a dot in the marker property name is kind of annoying as I can't write `e.customDimensions._MS.Example` in my Kusto queries (I have to use `customDimensions["_MS.Example"]`).

**Comment by SergeyKanzhelev** [comment]
@ohadschn makes sense. There are two reasons for this module:
- there is no ultimate unhandled exception handler for every platform. Every big product implements their own exception handling middleware or base page class that may not be Application Insights aware. So with the good filtering of first chance exceptions we may not need to even try to catch up with those handlers.
- show you a "Boneheaded" exception that was hidden as somebody tried to get rid of "Vexing" ones. It is quite often I see code that `try{}catch` the universe while in reality the goal was to not blow out for the specific well-known exception

As you said we definitely need to have a better visualization and filtering story there.

Thanks for feedback. Let me know if you have specific ideas on what else to collect to make filtering easier.


BTW, internal exceptions like ones from `PerfCounterCollector` are marked explicitly as internal. You can filter those out. Idea was to make you trust the module to get visibility in EVERY exception.

**Comment by ohadschn** [comment]
@SergeyKanzhelev thank you for explaining. The way I see it:
 1. In an ideal world, AI would have a product-specific SDK for every big product, at least for MS products. Exception handling is just one aspect, you have stuff like configuration (e.g. `UseApplicationInsights` in ASP.NET Core), auto-tracking interesting events (e.g. requests in ASP.NET), tracking relevant metrics/counters, knowing when to flush, etc. In the absence of a specific (complete) SDK, personally I would prefer plugging into these specifics hooks myself (e.g. flush on `OnCloseAsync` in the service fabric case). I suppose having first chance exceptions is better than nothing (though it might not be trivial understanding which was the exception that caused your crash), but I wonder if we're "educating" people to be lazy...
 2. That's why I said "properly written code" :) A developer who is careless enough to try/catch the universe (I'm stealing that expression BTW) is IMHO unlikely to sift through dozens (hundreds?) of exceptions and make sure they are all vexing / exogenous. Perhaps a smart query could be written, but I suspect this distinction cannot be made by static exception information alone (such as type).

I see what you mean about the `PerfCounterCollector` being marked internal (I assume you mean the `AI (Internal)` operation name), and that' a nice additional level of filtering. I could make the claim that it shouldn't even be collected at all (since you know for sure it is not user code) - at least not by default. BTW this might actually be a boneheaded exception. Specifically it is a `NullReferenceException`, is that expected? I can mail you the exception details if you'd like.

**Comment by SergeyKanzhelev** [comment]
Yes, it's boneheaded and we will fix it. For the not collecting it - we discussing options. Our ideal situation is when the windows counter `# exceptions thrown` matches the metrics collected by this telemetry module. It is a fine line between ALL exceptions and I do not trust this thing as it may not have collected what I need. And we want to build trust.

Besides the overall metrics and sifting thru exceptions - one of scenario is to show 1st chance exceptions in requests timeline. This will help troubleshooting requests even when this exception was handled. Specifically addressing the "bad code" situation. In the end of the day it may not be your bad code, it may be part of a library you cannot change.

Thanks for feedback, keep it coming!

**Comment by yzorg** [comment]
My team tried using FirstChanceExceptionStatisticsTelemetryModule.  We're also using Azure ServiceBus. 
 This combination logs thousands of `TimeoutException` per day.  The timeouts seem to be well understood when using ServiceBus TCP protocol (we have decided it is not worth switching to AMQP).

First, I strongly agree with @ohadschn's suggestion to track first-chance exceptions separately from the rest of the exception telemetry.  Let me quote his strong points:

> Most ... (are) swamped with irrelevant exceptions. Maybe a different firstChanceException schema object would be better. Also, these exceptions currently appear just like any "real" exception in the "Failures" blade, skewing the statistics along the way (e.g. "Total Server Exceptions by Problem ID").

Second, on @ohadschn's suggestion "... AI would have a product-specific SDK for every big product".  
Maybe this ServiceBus "flood" is low-hanging fruit, i.e. product-specific support in AppInsights?

Our team had initially approached FCE telemetry as if it was recommended, but after reading this thread it seems more like an experimental feature.  We debated earlier today, should we write custom telemetry filtering for our ServiceBus applications?  The chance of catching a rare "boneheaded" exceptions is so low I'm recommending we just disable FCE telemetry, at least until the out-of-box experience gets better.

**Comment by SergeyKanzhelev** [comment]
@yzorg I appreciate your feedback. It will be helpful if you can post couple examples of the most annoying exception to this thread. Also - do you have any custom handlers where you call `TrackException` for ones that you are interested in? If so - are those thrown by your code ("exogenous" or "vexing") or ones indicating total crash ("fatal")?
